### ISSUE TITLE
InfluxDB 2 support

### ISSUE THREAD

**Issue by weq** [markwragg/PowerShell-Influx#24]
Any plans or ETA for supporting InfluxDB 2 also? It is in only minor changes to the REST query for posting afaik.

**Comment by markwragg** [comment]
It's not something I was aware of but sounds doable. Feel free to fork and contribute if you feel able, if not link me to some details and i'll take a look when I have time.

**Comment by weq** [comment]
Would've done so already except that I have a broken hand but will give it a looksie when the cast is gone.

**Comment by markwragg** [comment]
Found the documentation for Influx DB 2 but it doesn't seem to detail how you write metrics via the API: 
https://v2.docs.influxdata.com/v2.0/get-started/

It seems to focus on either using Telegraf of having Influx go and pull metrics via scrapers. Do you know if there's any documentation for how you do writes via the API?

**Comment by R-Studio** [comment]
> Found the documentation for Influx DB 2 but it doesn't seem to detail how you write metrics via the API:
> https://v2.docs.influxdata.com/v2.0/get-started/
> 
> It seems to focus on either using Telegraf of having Influx go and pull metrics via scrapers. Do you know if there's any documentation for how you do writes via the API?

I think there is not a major change for writing data to an InfluxDB 2.x. InfluxDB 2.x also uses the **line protocol**.
The differences that I know are:
![image](https://user-images.githubusercontent.com/7498854/100754174-08a08200-33eb-11eb-8575-29f236b8077c.png)
- InfluxDB 2.x don't use the same DB and retention policy structure. That means in InfluxDB 2.x they only use buckets and buckets have a configured retention policy. To write data you only have to know the bucket name.

More in the documentation: https://docs.influxdata.com/influxdb/v2.0/write-data/

_I hope this helps. :-)_

**Comment by GWS65** [comment]
Hi

I have been using this library on an Arduino and it is suitable for V2 too. https://github.com/bonitoo-io/influxdb-client-arduino
Some examples show how to connect and maybe this info helps...

Would be nice to see that PowerShell-Flux will support the next generation
of Influx.

GWS65

**Comment by R-Studio** [comment]
@markwragg because I already forked your project I was not able to fork your project again for the InfluxDB 2.x support, so that's why I paste the Write-Influx.ps1 with InfluxDB 2.x support here in the comment:

```powershell
Function Write-Influx {
    <#
        .SYNOPSIS
            Writes data to Influx via the REST API.

        .DESCRIPTION
            Use to send data in to an Influx database by providing a hashtable of tags and values.

        .PARAMETER InputObject
            A metric object (generated by one of the Get-*Metric cmdlets from this module) which can be provided as pipeline input.

        .PARAMETER Measure
            The name of the measure to be updated or created.

        .PARAMETER Tags
            A hashtable of tag names and values.

        .PARAMETER Metrics
            A hashtable of metric names and values.

        .PARAMETER Timestamp
            Specify the exact date and time for the measure data point. If not specified the current date and time is used.

        .PARAMETER Server
            The URL and port for the Influx REST API. Default: 'http://localhost:8086'
        
        .PARAMETER Bucket
            The name of the Influx bucket/database to write to. (This is an InfluxDB v2.x Parameter)

        .PARAMETER Token
            The token to authenticate with InfluxDB. (This is an InfluxDB v2.x Parameter)

        .PARAMETER Organisation
            The name of the Influx organisation. (This is an InfluxDB v2.x Parameter)

        .PARAMETER Database
            The name of the Influx database to write to. (This is an InfluxDB v1.x Parameter)

        .PARAMETER Credential
            A PSCredential object with the username and password to use if the Influx server has authentication enabled. (This is an InfluxDB v1.x Parameter)
        
        .PARAMETER Bulk
            Switch: Use to have all metrics transmitted via a single connection to Influx.

        .PARAMETER ExcludeEmptyMetric
            Switch: Use to exclude null or empty metric values from being sent. Useful where a metric is initially created as an integer but then
            an empty or null instance of that metric would attempt to be sent as an empty string, resulting in a datatype conflict.

        .EXAMPLE
            Write-Influx -Measure WebServer -Tags @{Server='Host01'} -Metrics @{CPU=100; Memory=50} -Database Web -Server http://myinflux.local:8086
            
            Description
            -----------
            This command will submit the provided tag and metric data for a measure called 'WebServer' to a database called 'Web' 
            via the API endpoint 'http://myinflux.local:8086'
    #>
    [cmdletbinding(SupportsShouldProcess = $true, ConfirmImpact = 'Medium')]
    param (
        #Parameters for both InfluxDB versions
        [Parameter(ParameterSetName = 'MetricObject', Mandatory = $true, ValueFromPipeline = $True, Position = 0)]
        [Parameter(ParameterSetName = 'Influx_v1', ValueFromPipeline = $True)]
        [Parameter(ParameterSetName = 'Influx_v2', ValueFromPipeline = $True)]
        [PSTypeName('Metric')]
        [PSObject[]]
        $InputObject,

        [Parameter(ParameterSetName = 'Measure', Mandatory = $true, Position = 2)]
        [Parameter(ParameterSetName = 'Influx_v1')]
        [Parameter(ParameterSetName = 'Influx_v2')]
        [string]
        $Measure,

        [Parameter(ParameterSetName = 'Measure')]
        [Parameter(ParameterSetName = 'Influx_v1')]
        [Parameter(ParameterSetName = 'Influx_v2')]
        [hashtable]
        $Tags,
        
        [Parameter(ParameterSetName = 'Measure', Mandatory = $true)]
        [Parameter(ParameterSetName = 'Influx_v1')]
        [Parameter(ParameterSetName = 'Influx_v2')]
        [hashtable]
        $Metrics,

        [Parameter(ParameterSetName = 'Measure')]
        [Parameter(ParameterSetName = 'Influx_v1')]
        [Parameter(ParameterSetName = 'Influx_v2')]
        [datetime]
        $TimeStamp,
        
        [string]
        $Server = 'http://localhost:8086',

        [switch]
        $Bulk,

        [switch]
        $ExcludeEmptyMetric,

        #Parameters for InfluxDB v1.x
        [Parameter(ParameterSetName = 'Influx_v1', Mandatory = $true, Position = 1)]
        [switch]
        $InfluxDB_v1,
        
        [Parameter(ParameterSetName = 'Influx_v1', Mandatory = $true)]
        [string]
        $Database,

        [Parameter(ParameterSetName = 'Influx_v1', Mandatory = $true)]
        [pscredential]
        $Credential,

        #Parameters for InfluxDB v2.x
        [Parameter(ParameterSetName = 'Influx_v2', Mandatory = $true, Position = 1)]
        [switch]
        $InfluxDB_v2,

        [Parameter(ParameterSetName = 'Influx_v2', Mandatory = $true)]
        [string]
        $Organisation,

        [Parameter(ParameterSetName = 'Influx_v2', Mandatory = $true)]
        [string]
        $Bucket,

        [Parameter(ParameterSetName = 'Influx_v2', Mandatory = $true)]
        [string]
        $Token
    )

    Begin {
        if (($InfluxDB_v1) -and ($Credential)) {
            $Username = $Credential.UserName
            $Password = $Credential.GetNetworkCredential().Password

            $EncodedCreds = [System.Convert]::ToBase64String([System.Text.Encoding]::ASCII.GetBytes("$($Username):$($Password)"))

            $Headers = @{
                Authorization = "Basic $EncodedCreds"
            }
        }

        If ($InfluxDB_v1) {
            $URI = "$Server/write?&db=$Database"
        } else {
            $Headers = @{
                Authorization = "Token $Token"
            }
            
            $URI = "$Server/api/v2/write?org=$Organisation&bucket=$Bucket"
        }

        $BulkBody = @()
    }
    Process {
        if (-not $InputObject) {
            $InputObject = @{
                Measure = $Measure
                Metrics = $Metrics
                Tags = $Tags
                TimeStamp = $TimeStamp
            }
        }

        ForEach ($MetricObject in $InputObject) {
            if ($MetricObject.TimeStamp) {
                $timeStampNanoSecs = $MetricObject.Timestamp | ConvertTo-UnixTimeNanosecond
            }
            else {
                $null = $timeStampNanoSecs
            }

            if ($MetricObject.Tags) {
                $TagData = foreach ($Tag in $MetricObject.Tags.Keys) {
                    if ([string]::IsNullOrEmpty($MetricObject.Tags[$Tag])) {
                        Write-Warning "$Tag skipped as it's value was null or empty, which is not permitted by InfluxDB."
                    }
                    else {
                        "$($Tag | Out-InfluxEscapeString)=$($MetricObject.Tags[$Tag] | Out-InfluxEscapeString)"
                    }
                }
                $TagData = $TagData -Join ','
                $TagData = ",$TagData"
            }
        
            $Body = foreach ($Metric in $MetricObject.Metrics.Keys) {
            
                if ($ExcludeEmptyMetric -and [string]::IsNullOrEmpty($MetricObject.Metrics[$Metric])) {
                    Write-Verbose "$Metric skipped as -ExcludeEmptyMetric was specified and the value is null or empty."
                }
                Else {
                    if ($MetricObject.Metrics[$Metric] -isnot [ValueType]) { 
                        $MetricValue = '"' + $MetricObject.Metrics[$Metric] + '"'
                    }
                    else {
                        $MetricValue = $MetricObject.Metrics[$Metric] | Out-InfluxEscapeString
                    }
            
                    "$($MetricObject.Measure | Out-InfluxEscapeString)$TagData $($Metric | Out-InfluxEscapeString)=$MetricValue $timeStampNanoSecs"
                }            
            }
        
            if ($Body) {
                $Body = $Body -Join "`n"
            
                If ($Bulk) {
                    $BulkBody += $Body
                }
                Else {
                    if ($PSCmdlet.ShouldProcess($URI, $Body)) {
                        Invoke-RestMethod -Uri $URI -Method Post -Body $Body -Headers $Headers | Out-Null
                    }
                }
            
            }
        }
        
    }
    End {
        If ($Bulk) {
            $BulkBody = $BulkBody -Join "`n"
            
            if ($PSCmdlet.ShouldProcess($URI, $BulkBody)) {
                Invoke-RestMethod -Uri $URI -Method Post -Body $BulkBody -Headers $Headers | Out-Null
            }
        }
    }
}
```

**Comment by R-Studio** [comment]
@markwragg were you able to test my code?

**Comment by R-Studio** [comment]
@weq I add support for InfluxDB v2.x on my forked repo (https://github.com/R-Studio/PowerShell-Influx). I hope this helps.
_Notice: right now, I only add the support for using "Write-Influx.ps1" not for the UDP-Version._

@markwragg I create a pull request for the changes.

**Comment by markwragg** [comment]
Implemented via #35
